File: apps/api/services/routing/source_selector.py
"""Intent -> source_ids, filtered by tenant enabled integrations."""

from supabase import create_client
from config import settings

INTENT_TO_SOURCES: dict[str, list[str]] = {
    "crm": ["bitrix"],
    "documents": ["google_drive"],
    "spreadsheets": ["google_sheets"],
    "mixed": ["bitrix", "google_drive", "google_sheets"],
}


def get_sources_for_intent(tenant_id: str, intent_label: str) -> list[str]:
    """
    Return ordered source_ids for this intent that the tenant has enabled.
    Empty if no enabled integration for any of the mapped sources.
    """
    source_ids = INTENT_TO_SOURCES.get(intent_label)
    if not source_ids:
        return []
    if not settings.supabase_url or not settings.supabase_key:
        return []
    client = create_client(settings.supabase_url, settings.supabase_key)
    r = (
        client.table("integrations")
        .select("type")
        .eq("tenant_id", tenant_id)
        .eq("enabled", True)
        .in_("type", source_ids)
        .execute()
    )
    enabled_types = {row["type"] for row in (r.data or [])}
    return [s for s in source_ids if s in enabled_types]

File: apps/web/lib/server/routing/sourceSelector.ts
/**
 * Intent -> source_ids, filtered by tenant enabled integrations. Server-only.
 */

import { INTENT_TO_SOURCES, type IntentLabel } from "@repo/shared";
import { createServerSupabaseClient } from "../supabase-admin";

export async function getSourcesForIntent(
  tenantId: string,
  userId: string,
  intentLabel: string
): Promise<string[]> {
  const sourceIds = INTENT_TO_SOURCES[intentLabel as IntentLabel];
  if (!sourceIds || sourceIds.length === 0) return [];
  const hasServerSupabaseKey =
    !!process.env.SUPABASE_KEY || !!process.env.SUPABASE_SERVICE_ROLE_KEY;
  if (!process.env.SUPABASE_URL || !hasServerSupabaseKey) return [];
  const supabase = createServerSupabaseClient();
  const [tenantRes, userRes] = await Promise.all([
    supabase
      .from("integrations")
      .select("type")
      .eq("tenant_id", tenantId)
      .eq("enabled", true)
      .in("type", [...sourceIds]),
    supabase
      .from("user_integrations")
      .select("type")
      .eq("tenant_id", tenantId)
      .eq("user_id", userId)
      .eq("enabled", true)
      .in("type", [...sourceIds]),
  ]);

  const tenantEnabledTypes = new Set((tenantRes.data ?? []).map((r) => r.type as string));
  const userEnabledTypes = new Set((userRes.data ?? []).map((r) => r.type as string));

  // Tenant has priority: if tenant integration is enabled, it wins regardless of user overrides.
  return sourceIds.filter((id) => tenantEnabledTypes.has(id) || userEnabledTypes.has(id));
}
